bindings/rust/s2n-tls/src/testing.rs

```diff
@@ -325,3 +335,205 @@ pub fn poll_tls_pair_result(pair: &mut Pair<Harness, Harness>) -> Result<()> {
+            //             connections are moved out from the struct.
+            //
+            // cast 3: *const LocalDataBuffer as *mut c_void -> cast into the final *mut c_void required
+            //     SAFETY: serialized access is enforced by the interior RefCell, so it is safe to vend out
+            //             multiple mutable pointers to this item. We ensure this by casting back to an immutable
+            //             reference in the send and recv callbacks
+            conn.set_send_context(
+                send_ctx as &LocalDataBuffer as *const LocalDataBuffer as *mut c_void,
+            )?
+            .set_receive_context(
+                recv_ctx as &LocalDataBuffer as *const LocalDataBuffer as *mut c_void,
+            )?;
+        }
+        Ok(conn)
+    }
+
+    /// perform a TLS handshake between the connections
+    ///
+    /// This method will call `poll_negotiate` on each connection until both return
+    /// Ready(Ok) which indicates a successful handshake, or until one of the connections
+    /// returns Ready(Err) which indicates some fatal error.
+    pub fn handshake(&mut self) -> Result<(), error::Error> {
+        loop {
+            match (self.client.poll_negotiate(), self.server.poll_negotiate()) {
+                // if everything is finished and Ok, return Ok
+                (Poll::Ready(Ok(_)), Poll::Ready(Ok(_))) => return Ok(()),
+                // if there has been an error on the server
+                (_, Poll::Ready(Err(e))) => return Err(e),
+                // or an error on the client, then return the error
+                (Poll::Ready(Err(e)), _) => return Err(e),
+                _ => { /* not ready, poll again */ }
+            }
+        }
+    }
+
+    unsafe extern "C" fn send_cb(context: *mut c_void, data: *const u8, len: u32) -> c_int {
+        let context = &*(context as *const LocalDataBuffer);
+        let data = core::slice::from_raw_parts(data, len as _);
+        let bytes_written = context.borrow_mut().write(data).unwrap();
+        bytes_written as c_int
+    }
+
+    // Note: this callback will be invoked multiple times in the event that
+    // the byte-slices of the VecDeque are not contiguous (wrap around).
+    unsafe extern "C" fn recv_cb(context: *mut c_void, data: *mut u8, len: u32) -> c_int {
+        let context = &*(context as *const LocalDataBuffer);
+        let data = core::slice::from_raw_parts_mut(data, len as _);
+        match context.borrow_mut().read(data) {
+            Ok(len) => {
+                if len == 0 {
+                    // returning a length of 0 indicates a channel close (e.g. a
+                    // TCP Close) which would not be correct here. To just communicate
+                    // that there is no more data, we instead set the errno to
+                    // WouldBlock and return -1.
+                    errno::set_errno(errno::Errno(libc::EWOULDBLOCK));
+                    -1
```

Comment:
How would we indicated an unexpected end of stream? Is that just not testable with this? If so that's probably fine, I'm just checking my understanding.

Reply:
It's not (directly) testable with this harness. Communication over shared memory doesn't really have a semantic corollary for "unexpected end of stream" 🤔 .

If you wanted to write a unit test covering this, I think the easiest way would be to override the send/recv callbacks to some "mockable" variety which returns the errno/value you're interested in.